filters/StatsFilter.hpp

```diff
@@ -68,15 +68,52 @@ typedef std::vector<double> DataVector;
     double average() const
         { return M1; }
     double variance() const
-        { return M2/(m_cnt - 1.0); }
+        { return M2 / m_cnt; }
+    double sampleVariance() const
+        { return M2 / (m_cnt - 1.0); }
     double stddev() const
         { return std::sqrt(variance()); }
+    double sampleStddev() const
+        { return std::sqrt(sampleVariance()); }
     double skewness() const
-        { return (M2 && m_advanced) ?
-            std::sqrt(double(m_cnt)) * M3 / std::pow(M2, 1.5) : 0.0; }
+    {
+        if (!M2 || ! m_advanced)
+            return 0;
+        return std::sqrt(double(m_cnt)) * M3 / std::pow(M2, 1.5);
+    }
+    double sampleSkewness() const
+    {
+        if (M2 == 0 || m_cnt <= 2 || !m_advanced)
+            return 0.0;
+        double c(m_cnt);
+        return skewness() * std::sqrt(c) * std::sqrt(c - 1) / (c - 2);
+    }
     double kurtosis() const
-        { return (M2 && m_advanced) ?
-            double(m_cnt)*M4 / (M2*M2) - 3.0 : 0.0; }
+    {
+        if (M2 == 0 || !m_advanced)
+            return 0;
+        return double(m_cnt) * M4 / (M2 * M2);
+    }
+    double excessKurtosis() const
+    {
+        if (M2 == 0 || !m_advanced)
+            return 0;
+        return kurtosis() - 3;
+    }
+    double sampleKurtosis() const
+    {
+        if (M2 == 0 || m_cnt <= 3 || !m_advanced)
+            return 0;
+        double c(m_cnt);
+        return kurtosis() * (c + 1) * (c - 1) / ((c - 2) * (c - 3));
```

Comment:
I didn't go through the math, so I'll just ask, was it intentional to do `(c+1)*(c-1)`? I see that it is `(c-1)*(c-1)` in the sample excess kurtosis, so I'm guessing that once you expand all the terms there must be some cancellation that is going on.

Reply:
Yes.  Intentional.  The 'c' term is already included in the `kurtosis()` calculation.  I don't understand all the math.  `M2` is the numerator of the variance, so dividing by the square of the variance gets you two `c - 1` terms in the numerator.  One of those cancels with the `c - 1` that's in the denominator that comes up in the adjustment from the population statistic to the sample statistic.  Strangely, this finance site has some of the nicest and clearest formula: https://www.macroption.com/kurtosis-formula/

I've verified the calculations against google sheets and that against other websites (which is how I figured out that what people are selling as "kurtosis" is actually "sample excess kurtosis", so I'm pretty confident that they're correct at this point.